### DISCUSSION TITLE
Miryoku firmware in Python

### DISCUSSION THREAD

**Discussion by akaenner** [manna-harbour/miryoku#58] [Show and tell]
Hi,

I'm a big fan of the Miryoku layout and I created my own keyboard and firmware using CircuitPython to implement something similar. Just because I love systems without big dependencies. The firmware is fully self contained and does not need a development environment.

The firmware: https://github.com/akaenner/picosplit

The keyboard: https://kaenner.de/picosplit

**Comment by manna-harbour** [comment]
Thanks very much for sharing your work here!  I hope Show and Tell will be used to share anything related to Miryoku, so this is an excellent start to that!

The keyboard looks great, and the build guide is wonderful!  I've long wanted to build a DM with a smoothly contoured case.  And it's nice that the MCUs have enough I/O to wire the switches directly.  I also love the use of magnets.  I use [magnetic tent legs](https://www.reddit.com/r/MechanicalKeyboards/comments/ebf228/diy_magnetic_tent_legs/), and am planning a magnetic alternative to the tripod socket used [here](https://www.reddit.com/r/ErgoMechKeyboards/comments/ibvzr8/arm_mount/).  

The firmware is great too!  I think I've seen other Circuit Python based firmware, but maybe not with mod tap support suitable for home row mods.  It'd be pretty straightforward to generate your `layout.js` from Miryoku Babel and a makefile, as is done in Miryoku KMonad.  That wouldn't be a good match for your goals of having the firmware be self-contained though!  But since the Miryoku Babel tangle scrips are already python, a more interesting approach would be to integrate those plus a python version of the tables, then a simple config file could specify the alternative layout options, and the keymap could be generated at run time from solely within the firmware.

**Reply by akaenner** [reply]
Thank you for the kind words. A special feature might be that with the PicoSplit firmware you don't have to pay so much attention to which pin you connect which button to. The assignment to the layout is determined later by a static web page, which contains a Javascript program. You just have to press all the keys once and the layout fits afterwards. If you wire a keyboard by hand, this can be an advantage. The Raspberry Pi Pico also presents itself as a USB memory drive, where all the configuration files are located.
